src/gc/gc.cpp

```diff
@@ -15882,14 +15882,42 @@ start_no_gc_region_status gc_heap::prepare_for_no_gc_region (uint64_t total_size
     save_data_for_no_gc();
     settings.pause_mode = pause_no_gc;
     current_no_gc_region_info.start_status = start_no_gc_success;
-    
+
     size_t allocation_no_gc_loh = 0;
     size_t allocation_no_gc_soh = 0;
     size_t size_per_heap = 0;
 
+    // requested sizes of 0 make no sense.
+    if (total_size == 0)
+    {
+        status = start_no_gc_too_large;
+        goto done;
+    }
+
+    if (loh_size_known)
+    {
+        if (loh_size == 0)
+        {
+            status = start_no_gc_too_large;
+            goto done;
+        }
+
+        // According to the documentation, TryStartNoGCRegion must fail if
+        // it can't allocate total_size - loh_size for the SOH. This only makes
+        // any kind of sense if total_size > loh_size.
+        if (total_size < loh_size)
+        {
+            status = start_no_gc_too_large;
+            goto done;
+        }
+    }
+
+    assert(total_size != 0);
     total_size = (size_t)((float)total_size * 1.05);
```

Comment:
This isn't strictly part of your change, but I looked at the surrounding code and I'm uncomfortable with the casting going on. `total_size` comes in as a uint64_t, but then we do these float/size_t casts. The loss of precision and the use of `size_t` can lead to unexpected behavior on x86.

Reply:
Very true. It's unfortunate that we are stuck with managed `long` parameters on `GC.TryStartNoGCRegion`.  We should probably bail if:

* total_size > `UINTPTR_MAX`
* loh_size > `UINTPTR_MAX`
* total_size is less than the original total_size, after the float mul and cast
* loh_size is less than the original total_size, after the float mul and cast

Reply:
Ugh... I just looked at the spec and downcasting from a float that won't fit into a `size_t` is undefined behavior, so I guess it won't be fine.